File: frameworks/shared/typescript/middleware.ts
/**
 * Middleware for automatic health check endpoint handling.
 *
 * Provides middleware factories for Express and Fastify that automatically
 * handle health check requests without requiring manual route definitions.
 */

import { Request, Response, NextFunction } from 'express';
import { FastifyRequest, FastifyReply } from 'fastify';
import { HealthCheckManager } from './health-check';
import { getHttpStatusCode, ProbeType } from './types';

/**
 * Express middleware for health check endpoints.
 *
 * Usage:
 *   import { expressHealthCheckMiddleware } from '@shared/middleware';
 *
 *   const healthManager = new HealthCheckManager({ ... });
 *   app.use(expressHealthCheckMiddleware(healthManager));
 */
export function expressHealthCheckMiddleware(healthManager: HealthCheckManager) {
  return async (req: Request, res: Response, next: NextFunction) => {
    const path = req.path;

    // Check if this is a health check endpoint
    if (req.method === 'GET' && path.startsWith('/health')) {
      let probeType: ProbeType;

      if (path === '/health') {
        probeType = ProbeType.READINESS; // Default to readiness
      } else if (path === '/health/live') {
        probeType = ProbeType.LIVENESS;
      } else if (path === '/health/ready') {
        probeType = ProbeType.READINESS;
      } else if (path === '/health/startup') {
        probeType = ProbeType.STARTUP;
      } else {
        // Unknown health endpoint
        return res.status(404).json({ error: 'Unknown health endpoint' });
      }

      try {
        // Execute health check
        const result = await healthManager.probe(probeType);
        const statusCode = getHttpStatusCode(result.status, result.probe_type);

        return res.status(statusCode).json(result);
      } catch (error: any) {
        return res.status(503).json({
          status: 'down',
          error: `Health check error: ${error.message}`,
        });
      }
    }

    // Not a health check, pass to next middleware
    next();
  };
}

/**
 * Fastify plugin for health check endpoints.
 *
 * Usage:
 *   import { fastifyHealthCheckPlugin } from '@shared/middleware';
 *
 *   const healthManager = new HealthCheckManager({ ... });
 *   app.register(fastifyHealthCheckPlugin, { healthManager });
 */
export async function fastifyHealthCheckPlugin(
  fastify: any,
  options: { healthManager: HealthCheckManager }
) {
  const { healthManager } = options;

  // Register health check routes
  fastify.get('/health', async (request: FastifyRequest, reply: FastifyReply) => {
    try {
      const result = await healthManager.probe('readiness');
      const statusCode = getHttpStatusCode(result.status, result.probe_type);
      reply.code(statusCode).send(result);
    } catch (error: any) {
      reply.code(503).send({
        status: 'down',
        error: `Health check error: ${error.message}`,
      });
    }
  });

  fastify.get('/health/live', async (request: FastifyRequest, reply: FastifyReply) => {
    try {
      const result = await healthManager.probe('liveness');
      const statusCode = getHttpStatusCode(result.status, result.probe_type);
      reply.code(statusCode).send(result);
    } catch (error: any) {
      reply.code(503).send({
        status: 'down',
        error: `Health check error: ${error.message}`,
      });
    }
  });

  fastify.get('/health/ready', async (request: FastifyRequest, reply: FastifyReply) => {
    try {
      const result = await healthManager.probe('readiness');
      const statusCode = getHttpStatusCode(result.status, result.probe_type);
      reply.code(statusCode).send(result);
    } catch (error: any) {
      reply.code(503).send({
        status: 'down',
        error: `Health check error: ${error.message}`,
      });
    }
  });

  fastify.get('/health/startup', async (request: FastifyRequest, reply: FastifyReply) => {
    try {
      const result = await healthManager.probe('startup');
      const statusCode = getHttpStatusCode(result.status, result.probe_type);
      reply.code(statusCode).send(result);
    } catch (error: any) {
      reply.code(503).send({
        status: 'down',
        error: `Health check error: ${error.message}`,
      });
    }
  });
}

File: frameworks/shared/go/middleware.go
package healthcheck

import (
	"encoding/json"
	"net/http"
)

// HTTPHandler returns an http.HandlerFunc that handles health check requests.
//
// Usage:
//   healthManager := healthcheck.New(healthcheck.Config{ ... })
//   http.HandleFunc("/health", healthcheck.HTTPHandler(healthManager, "readiness"))
//   http.HandleFunc("/health/live", healthcheck.HTTPHandler(healthManager, "liveness"))
func HTTPHandler(manager *HealthCheckManager, probeType string) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet {
			http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
			return
		}

		result, err := manager.Probe(probeType)
		if err != nil {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusServiceUnavailable)
			json.NewEncoder(w).Encode(map[string]interface{}{
				"status": "down",
				"error":  err.Error(),
			})
			return
		}

		statusCode := result.GetHTTPStatusCode()
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(statusCode)
		json.NewEncoder(w).Encode(result)
	}
}

// RegisterRoutes registers health check routes with the default HTTP mux.
//
// Usage:
//   healthManager := healthcheck.New(healthcheck.Config{ ... })
//   healthcheck.RegisterRoutes(http.DefaultServeMux, healthManager)
func RegisterRoutes(mux *http.ServeMux, manager *HealthCheckManager) {
	mux.HandleFunc("/health", HTTPHandler(manager, "readiness"))
	mux.HandleFunc("/health/live", HTTPHandler(manager, "liveness"))
	mux.HandleFunc("/health/ready", HTTPHandler(manager, "readiness"))
	mux.HandleFunc("/health/startup", HTTPHandler(manager, "startup"))
}

// GinMiddleware returns a Gin middleware function for health checks.
//
// Usage (Gin framework):
//   import "github.com/gin-gonic/gin"
//
//   healthManager := healthcheck.New(healthcheck.Config{ ... })
//   router := gin.Default()
//   router.GET("/health", healthcheck.GinHandler(healthManager, "readiness"))
func GinHandler(manager *HealthCheckManager, probeType string) interface{} {
	// Return generic interface{} to avoid importing gin
	// Actual usage will type-assert to gin.HandlerFunc
	return func(c interface{}) {
		// Type assertion would happen in actual implementation
		// This is a placeholder showing the pattern
	}
}

// EchoMiddleware returns an Echo middleware function for health checks.
//
// Usage (Echo framework):
//   import "github.com/labstack/echo/v4"
//
//   healthManager := healthcheck.New(healthcheck.Config{ ... })
//   e := echo.New()
//   e.GET("/health", healthcheck.EchoHandler(healthManager, "readiness"))
func EchoHandler(manager *HealthCheckManager, probeType string) interface{} {
	// Return generic interface{} to avoid importing echo
	// Actual usage will type-assert to echo.HandlerFunc
	return func(c interface{}) error {
		// Type assertion would happen in actual implementation
		// This is a placeholder showing the pattern
		return nil
	}
}
